nntrainer/layers/mol_attention_layer.cpp

```diff
@@ -72,34 +137,55 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
   Tensor &state = context.getInput(wt_idx[AttentionParams::state]);
 
   Tensor &output = context.getOutput(SINGLE_INOUT_IDX);
-  Tensor &mlp_w = context.getWeight(wt_idx[AttentionParams::mlp_w]);
-  Tensor &mlp_proj_w = context.getWeight(wt_idx[AttentionParams::mlp_proj_w]);
+  Tensor &fc_w = context.getWeight(wt_idx[AttentionParams::fc_w]);
+  Tensor &fc_bias = context.getWeight(wt_idx[AttentionParams::fc_bias]);
+  Tensor &fc_proj_w = context.getWeight(wt_idx[AttentionParams::fc_proj_w]);
+  Tensor &fc_out = context.getTensor(wt_idx[AttentionParams::fc_out]);
+  Tensor &fc_tanh = context.getTensor(wt_idx[AttentionParams::fc_tanh]);
+  Tensor &fc_proj_out = context.getTensor(wt_idx[AttentionParams::fc_proj_out]);
+  Tensor &scores = context.getTensor(wt_idx[AttentionParams::scores]);
+  Tensor &prob = context.getTensor(wt_idx[AttentionParams::prob]);
+  Tensor &prob_left = context.getTensor(wt_idx[AttentionParams::prob_left]);
+  Tensor &prob_right = context.getTensor(wt_idx[AttentionParams::prob_right]);
+  Tensor &u_neg_div = context.getTensor(wt_idx[AttentionParams::u_neg_div]);
+  Tensor &u_pos_div = context.getTensor(wt_idx[AttentionParams::u_pos_div]);
 
   const TensorDim &input_dim = query.getDim();
   unsigned int batch = input_dim.batch();
   auto mol_k = std::get<props::MoL_K>(mol_props).get();
 
-  Tensor mlp_out = query.dot(mlp_w);
+  /** reset helper state */
+  helper_exec = false;
+
+  fc_out = query.dot(fc_w);
+  fc_out.add_i(fc_bias);
 
-  Tensor mlp_tanh;
-  tanh.run_fn(mlp_out, mlp_tanh);
+  tanh.run_fn(fc_out, fc_tanh);
 
-  Tensor mlp_proj_out = mlp_tanh.dot(mlp_proj_w);
+  fc_proj_out = fc_tanh.dot(fc_proj_w);
 
-  /** @note kappa_hat, beta_hat are strided */
-  Tensor kappa_hat =
-    mlp_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, 0, false);
-  Tensor beta_hat =
-    mlp_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, mol_k, false);
-  Tensor alpha_hat;
-  alpha_hat.copy_with_stride(
-    mlp_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, mol_k * 2, false));
+  Tensor kappa_src, beta_src, alpha_src;
+  kappa_src.copy_with_stride(
+    fc_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, 0, false));
+  beta_src.copy_with_stride(
+    fc_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, mol_k, false));
+  alpha_src.copy_with_stride(
+    fc_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, mol_k * 2, false));
 
-  Tensor kappa = kappa_hat.apply(static_cast<float (*)(float)>(&std::exp));
-  Tensor beta = beta_hat.apply(static_cast<float (*)(float)>(&std::exp));
+  kappa_src.apply_i(&expf);
+  beta_src.apply_i(&expf);
+  Tensor kappa = kappa_src;
+  Tensor beta = beta_src;
 
   Tensor alpha;
-  softmax.run_fn(alpha_hat, alpha);
+  softmax.run_fn(alpha_src, alpha);
```

Comment:
Doesn't this throw considering below?
```
Tensor &ActiFunc::softmax(Tensor const &t, Tensor &output) {
  /**
   * shiftx_logit = logit - max_batch(logit)
   * softmax = exp(shiftx_logit) / (sum(exp(shiftx_logit)))
   *
   * @note softmax is applied on the last dimension
   */
  unsigned int fixed_dim = t.getDim().getDataLen() / t.width();
  float *dp;
  float *rp;

  /** TODO: support strided operations */
  if (t.size() == output.size() && t.getStrides() != output.getStrides())
    throw std::invalid_argument(
      "Softmax does not support operating on strided tensors");      ```

Reply:
because alpha_src is not a shared strided tensor but a contiguous tensor with data copied in it.

Reply:
Let me know if I didnt answer your question.

Reply:
I wonder alpha.size() is empty?

Reply:
alpha is indeed empty, but it does not throw as empty tensor is not strided

Reply:
Ah missed that part. Thanks :)